Meter.c

```diff
@@ -266,89 +317,610 @@ static void BarMeterMode_draw(Meter* this, int x, int y, int w) {
 
 /* ---------- GraphMeterMode ---------- */
 
-#ifdef HAVE_LIBNCURSESW
+static void GraphMeterMode_createGraphData(Meter* this, unsigned int numRecords) {
+   assert(this->drawData == NULL);
+
+   GraphData* data = xCalloc(1, sizeof(GraphData));
+   data->numRecords = numRecords;
+
+   bool isOneColorGraph = GRAPH_HEIGHT >= 2 && Meter_maxItems(this) <= 1;
+
+   // recordNumCells determines number of cells needed to allocate for a graph
+   // data record. A record corresponds to a terminal column when drawn.
+   size_t recordNumCells;
+   if (isOneColorGraph) {
+      // Index 0: log2 of scale
+      // Index 1: Number of dots total
+      recordNumCells = 2;
+   } else if (this->total > 0.0) {
+      // Percent graph
+      // Index 0..GRAPH_HEIGHT: Color cells
+      recordNumCells = GRAPH_HEIGHT;
+   } else {
+      // Non-percentage & dynamic scale
+      // Index 0: log2 of scale
+      // Index 1..(2*GRAPH_HEIGHT-1): Color cells
+      recordNumCells = 2 * GRAPH_HEIGHT;
+   }
+   assert(recordNumCells <= UINT16_MAX);
+   data->recordNumCells = (uint16_t)recordNumCells;
+
+   GraphColorCell* cells = xCalloc(numRecords * recordNumCells, sizeof(GraphColorCell));
+   if (!isOneColorGraph) {
+      for (size_t i = 0; i < numRecords * recordNumCells; i++) {
+         if (this->total > 0.0 || i % recordNumCells > 0) {
+            cells[i].cell.itemIndex = UINT8_MAX;
+         }
+      }
+   }
+   data->cells = cells;
+   this->drawData = data;
+}
 
-#define PIXPERROW_UTF8 4
-static const char* const GraphMeterMode_dotsUtf8[] = {
-   /*00*/" ", /*01*/"⢀", /*02*/"⢠", /*03*/"⢰", /*04*/ "⢸",
-   /*10*/"⡀", /*11*/"⣀", /*12*/"⣠", /*13*/"⣰", /*14*/ "⣸",
-   /*20*/"⡄", /*21*/"⣄", /*22*/"⣤", /*23*/"⣴", /*24*/ "⣼",
-   /*30*/"⡆", /*31*/"⣆", /*32*/"⣦", /*33*/"⣶", /*34*/ "⣾",
-   /*40*/"⡇", /*41*/"⣇", /*42*/"⣧", /*43*/"⣷", /*44*/ "⣿"
-};
+static void GraphMeterMode_computeColors(Meter* this, GraphColorCell* record, int numDots, double scaledTotal, int scaleExp, bool isPercentGraph) {
+   // If there is a "top cell" which won't be completely filled, determine
+   // its color first.
+   assert(numDots > 0 && numDots <= (GRAPH_HEIGHT * 8));
+
+   int topCell = (numDots - 1) / 8;
+   uint8_t dotAlignment = 2;
+   unsigned int blanksAtTopCell = (8 + topCell * 8 - numDots) / dotAlignment * dotAlignment;
+   uint8_t topCellItem = this->curItems - 1;
+   double topCellArea = 0.0;
+   if (blanksAtTopCell > 0 ||
+      (!isPercentGraph &&
+      topCell == ((GRAPH_HEIGHT - 1) >> scaleExp) &&
+      topCell % 2 == 0)) {
+      topCellArea = (8 - blanksAtTopCell) / 8.0;
+      double sum = 0.0;
+      double bottom = 0.0;
+      double maxArea = 0.0;
+      for (uint8_t i = 0; i < this->curItems && sum < DBL_MAX; i++) {
+         // Assuming IEEE 754 semantics, the "!(>)" comparison includes NaN.
+         if (!(this->values[i] > 0.0)) // Skip negative and NaN values
+            continue;
+         sum += this->values[i];
```

Comment:
Some additional space doesn't hurt …
```suggestion

         sum += this->values[i];
```